core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -294,18 +277,20 @@ class ProducerIdMapping(val config: LogConfig,
       throw new IllegalArgumentException("Invalid PID passed to update")
     val entry = appendInfo.lastEntry
     pidMap.put(appendInfo.pid, entry)
-    lastMapOffset = entry.lastOffset + 1
+  }
+
+  def updateLastOffset(lastOffset: Long): Unit = {
+    lastMapOffset = lastOffset
   }
 
   /**
    * Load a previously stored PID entry into the cache. Ignore the entry if the timestamp is older
    * than the current time minus the PID expiration time (i.e. if the PID has expired).
    */
   def load(pid: Long, entry: ProducerIdEntry, currentTimeMs: Long) {
-    if (pid != RecordBatch.NO_PRODUCER_ID && currentTimeMs - entry.timestamp < maxPidExpirationMs) {
+    if (pid != RecordBatch.NO_PRODUCER_ID && currentTimeMs - entry.timestamp < maxPidExpirationMs)
       pidMap.put(pid, entry)
-      lastMapOffset = entry.lastOffset + 1
-    }
+    lastMapOffset = entry.lastOffset + 1
```

Comment:
Ack. This may have been my first attempt to fix the problem which I later forgot to clean up.